Update mongodb dashboard JSON

williamlardier opened the pull request:
Issue: ZENKO-5003

The scalgrafanalib integration will be done in ZENKO-5010

## updated Jul 21 11pm

<img width="1920" height="6642" alt="screencapture-10-160-103-37-8443-grafana-d-besmvgi5ovd34d-mongodb57-2025-07-21-23_07_06" src="https://github.com/user-attachments/assets/4a4b6716-43d6-4ec9-8841-5bb69e1bf84f" />

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -64,194 +30,304 @@
         }
       ],
-      "title": "Uptime",
+      "title": "MongoDB services state",
```
the text does not line up, does not look very polished. I think it would be better to use 3 small Stat panel, for each service, in the first row similar to all other dashboards (and `Version` can be moved to the next row, left of `smallest olog window)

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
-        "align": false,
-        "alignLevel": null
-      }
+      "title": "Number of documents per shard",
```
"per-shard" does not scale well, and having multiple time series in a Stats panel is not very usable in general (un-clear legend...)
May be ok for a first iteration, but will need to be reworked when we migrate the dashboard to python.
→ please add a checklist item or line in description there to remember to change it. 

Same for index-size panel.

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
-        "alignLevel": null
-      }
+      "title": "Number of documents per shard",
+      "type": "stat"
```
for the overview, I think we are missing some indication of the "Activity" (number of cursors? number of operations?...)

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+        }
+      ],
+      "title": "Shard balancing",
+      "type": "timeseries"
```
since this represents a boolean, I don't think a time series is the most appropriate: generally, a timeline panel may be more appropriate?

also there is no legend, which makes it hard to fully understand the panel.
→ either we should aggregate (i.e. "% of shards which are balanced") or we should have one state/timeline per series

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -64,194 +30,304 @@
+      "datasource": {
+        "default": true,
+        "type": "prometheus",
+        "uid": "metalk8s-prometheus"
```
should not be hard-coded, but template variable for zkop (like in the other dashboards)

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+            "uid": "metalk8s-prometheus"
+          },
+          "editorMode": "code",
+          "expr": "avg(mongodb_ss_metrics_cursor_open{rs_nm=~\"$rs_nm\"})",
```
* missing unit in panel
* what is the point of average? average number of cursor per mongos instance?

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+        }
+      ],
+      "title": "Active cursors",
+      "type": "stat"
```
best to use a timeseries, so that values can be explored (Stats should be used for "small" indicators, not full graphs). Also we don't really care about the overall value, but may be interested in min/max/avg...

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+          "refId": "B"
+        }
+      ],
+      "title": "Operations done by shard servers",
```
"Data operations" or "Shard operations" ?

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+          "step": 300
+        }
+      ],
+      "title": "Command Operations",
```
"Operations by command" ?

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+            "uid": "metalk8s-prometheus"
+          },
+          "editorMode": "code",
+          "expr": "sum by (rs_nm, legacy_op_type) (rate(mongodb_ss_opcountersRepl{rs_nm=~\"$rs_nm\", rs_state=\"1\"}[$__rate_interval]))",
```
should not keep `rs_nm` label, it creates too many dimensions in the panel : either aggregate it to have the overall operations, or have separate panels for each.
(even if we aggregate, we may however filter the metrics like you do, based on the users' selected filter in the dashboard)

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+            "uid": "metalk8s-prometheus"
+          },
+          "editorMode": "code",
+          "expr": "sum by(job, op_type) (\n  rate(mongodb_mongod_op_latencies_latency_total{job=~\".*shards.*\"}[$__rate_interval])\n) /\nsum by(job, op_type) (\n  rate(mongodb_mongod_op_latencies_ops_total{job=~\".*shards.*\"}[$__rate_interval])\n)",
```
- it seems we have a single `job` : should we not aggregate by "rs_nm" instead?
- keep only one dimension to display the data: either by rs_nm (each shard...) or by kind of operation

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+            "uid": "metalk8s-prometheus"
+          },
+          "editorMode": "code",
+          "expr": "max_over_time(mongodb_mongod_replset_member_replication_lag{job=~\"zenko/data-db-mongodb-sharded-shard.*-data\"}[$interval]) > 0 ",
```
should it not be `last_over_time` ?

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+        "type": "prometheus",
+        "uid": "metalk8s-prometheus"
+      },
+      "description": "Average latency of operations (classified by read, write, or (other) command)",
```
description does not match

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+            "min"
+          ],
+          "displayMode": "table",
+          "placement": "bottom",
```
labels are short and panels quite wide → best to move legend to the right

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+            "uid": "metalk8s-prometheus"
+          },
+          "editorMode": "code",
+          "expr": "avg by (op_type) (rate(mongodb_ss_opLatencies_latency{job=~\"zenko/data-db-mongodb-sharded-shard.*-data\"}[$__rate_interval]) / (rate(mongodb_ss_opLatencies_ops{job=~\"zenko/data-db-mongodb-sharded-shard.*-data\"}[$__rate_interval]) > 0) or irate(mongodb_ss_opLatencies_latency{job=~\"zenko/data-db-mongodb-sharded-shard.*-data\"}[$__rate_interval]) / (irate(mongodb_ss_opLatencie ...
```
is it really relevant to use `rate() or irate()` ? There are clearly documented reason to use either rate or irate, so we should just pick the appropriate one.

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+            "min"
+          ],
+          "displayMode": "table",
+          "placement": "bottom",
```
same, may be better aligned right

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+            "min"
+          ],
+          "displayMode": "table",
+          "placement": "bottom",
```
shouljd be right-aligned

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+            "min"
+          ],
+          "displayMode": "table",
+          "placement": "bottom",
```
place on the right

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
-        "shared": true,
-        "sort": 0,
-        "value_type": "individual"
+      "title": "Query Efficiency",
```
what does "efficiency" mean here?
the legend does not really help ("scanned / returned"), and there is no unit (for an efficiency, we would expect some sort of percentage or clear ratio)

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+          "expr": "(\n   rate(mongodb_ss_wt_cache_pages_requested_from_the_cache{rs_state=\"1\",rs_nm=~\"$rs_nm\"}[$__rate_interval]) - \n   rate(mongodb_ss_wt_cache_pages_read_into_cache{rs_state=\"1\",rs_nm=~\"$rs_nm\"}[$__rate_interval])) / \n   rate(mongodb_ss_wt_cache_pages_requested_from_the_cache{rs_state=\"1\",rs_nm=~\"$rs_nm\"}[$__rate_interval]\n)",
+          "hide": false,
+          "instant": false,
+          "legendFormat": "Cache hit/miss ratio for {{ pod }}",
```
redundant with graph title
```suggestion
          "legendFormat": "{{ pod }}",
```

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+          },
+          "editorMode": "code",
+          "exemplar": false,
+          "expr": "sum by (rs_nm, job) (\n  (\n    rate(mongodb_sys_cpu_user_ms{job=\"zenko/data-db-mongodb-sharded-shards-data\", rs_nm=~\"$rs_nm\"}[$__rate_interval]) +\n    rate(mongodb_sys_cpu_iowait_ms{job=\"zenko/data-db-mongodb-sharded-shards-data\", rs_nm=~\"$rs_nm\"}[$__rate_interval]) +\n    rate(mongodb_sys_cpu_system_ms{job=\"zenko/data-db-mongodb-sharded-shards-data\", rs_nm=~\"$rs_n ...
```
since this is a "stacked" graph, the series must be independent: so there should not be a "total" series

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+      "options": {
+        "legend": {
+          "calcs": [],
+          "displayMode": "hidden",
```
hard to read with no legend

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+            "axisPlacement": "auto",
+            "barAlignment": 0,
+            "barWidthFactor": 1,
+            "drawStyle": "bars",
```
why bars ? it does not seem to improve on the default line, but looks a bit out of place

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
-        "align": false,
-        "alignLevel": null
-      }
+      "title": "Stacked CPU usage",
```
there is a slight inconsistency here: the other panels document each pod's resource use, except this one which breaks the total usage by "kind of cpu usage"s
→ IMO what we want to know if each instance's CPU usage ; breaking down the "cause" of CPU usage can be done by going to k8s cpu usage dashboard

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
-        "show": true,
-        "values": []
+      "id": 1625,
+      "options": {
```
what is the unit ? For _usage_ I'd except a percentage maybe? Unless it actually is a _load_?

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
-        "align": false,
-        "alignLevel": null
-      }
+      "title": "Memory Usage",
```
too many time series IMHO : either aggregate resident+virtual, or duplicate panel
(same for network I/O)

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+            "min"
+          ],
+          "displayMode": "table",
+          "placement": "bottom",
```
should be placed right
(same for `Network Traffic` panel)

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+          "expr": "sum(rate(container_network_receive_bytes_total{pod=\"data-db-mongodb-sharded-shard0-data-1\"}[$interval]) or \nirate(container_network_receive_bytes_total{pod=\"data-db-mongodb-sharded-shard0-data-1\"}[5m])) or\nsum(rate(node_network_receive_bytes_total{pod=\"data-db-mongodb-sharded-shard0-data-1\"}[$interval])) or \nsum(irate(node_network_receive_bytes_total{pod=\"data-db-mong ...
+          "hide": false,
+          "instant": false,
+          "legendFormat": "Inbound (Shard data 1)",
```
- redundant with Network I/O
- too many axis (pod + inbound/outbound), like other panels
- should not hard-code pods and legend, just keep a label in the serie and let Grafana create display each series
- choose between rate or irate, no point making query more complex

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+      "id": 1593,
       "options": {
-        "alertThreshold": true
+        "legend": {
```
missing unit: ops/second (same for eviction threads removed)

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
-        "shared": false,
-        "sort": 0,
-        "value_type": "individual"
+      "title": "Number of threads for eviction",
```
```suggestion
      "title": "Number of active eviction threads",
```

it is useful to have both the "value" and its evolution (created/removed) ?

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+            "uid": "metalk8s-prometheus"
+          },
+          "editorMode": "code",
+          "expr": "avg by (type) ((max_over_time(mongodb_mongod_wiredtiger_concurrent_transactions_total_tickets{txn_rw=\"read\"}[$interval]) or\nmax_over_time(mongodb_mongod_wiredtiger_concurrent_transactions_total_tickets{txn_rw=\"read\"}[5m])) -\n(max_over_time(mongodb_mongod_wiredtiger_concurrent_transactions_out_tickets{txn_rw=\"read\"}[$interval]) or\nmax_over_time(mongodb_mongod_wiredtiger ...
```
instead of doing 2 computations (over interval or 5m), it is best to use `$rate_interval`

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+          "editorMode": "code",
+          "expr": "sum by (txn_rw) (mongodb_mongod_wiredtiger_concurrent_transactions_available_tickets{job=~\"zenko/data-db-mongodb-sharded-shard.*-data\"})",
+          "instant": false,
+          "legendFormat": "{{pod}}  ({{txn_rw}})",
```
`pod` not defined in the serie

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
-        "align": false,
-        "alignLevel": null
-      }
+      "title": "Concurrency tickets",
```
```suggestion
      "title": "Concurrency tickets by pod",
```

francoisferrand replied:
+ too many series here (pod x 2), split graph or aggregate (same for Wiretiger cache I/O and WirdTiger Cache panels)

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
-        "align": false,
-        "alignLevel": null
-      }
+      "title": "WiredTiger Concurrency Tickets Available",
```
Wiretiger is redundant here, this is the title of the section
```suggestion
      "title": "Available Concurrency Tickets",
```

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+            "properties": [
+              {
+                "id": "custom.transform",
+                "value": "negative-Y"
```
this is very weird: it shows a negative graph for a number of tickets (quite weird already), which does not match the values in the table...
→ if you want to see the graph separately, best to use 2 panels

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
-        "align": false,
-        "alignLevel": null
-      }
+      "title": "WiredTiger Cache",
```
redundant with section title
```suggestion
      "title": "Cache Size",
```

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
-        "shared": false,
-        "sort": 0,
-        "value_type": "individual"
+      "title": "Eviction of pages",
```
`Page Eviction` or `Evicted Pages` ?
Also missing unit (ops/second?)

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
+            "uid": "metalk8s-prometheus"
+          },
+          "editorMode": "code",
+          "expr": "rate(mongodb_ss_wt_cache_bytes_written_from_cache{job=~\"zenko/data-db-mongodb-sharded-shard.*-data\", rs_nm=~\"$rs_nm\"}[$__rate_interval])",
```
should not hard-code `zenko/data-db-mongodb-...`, but use template and/or variable parameters (${job}, etc...)

francoisferrand commented on monitoring/mongodb/dashboard.json:
```diff
@@ -265,1465 +341,5690 @@
-        "align": false,
-        "alignLevel": null
-      }
+      "title": "WiredTiger Cache I/O",
```
wiredtiger redundant with section
```suggestion
      "title": "Cache I/O",
```

bert-e commented:
# Hello williamlardier,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/zenko/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Conflict




There is a conflict between your branch `feature/ZENKO-5003-mongodb-dashboard` and the
destination branch `development/2.12`.

Please resolve the conflict on **the feature branch** (`feature/ZENKO-5003-mongodb-dashboard`).

```sh
git fetch && \
git checkout origin/feature/ZENKO-5003-mongodb-dashboard && \
git merge origin/development/2.12
```

Resolve merge conflicts and commit

```sh
git push origin HEAD:feature/ZENKO-5003-mongodb-dashboard
```

Seragonia committed:
Redo of the MongoDB dashboard

  The current dashboard proved to be useles in production, not useful in case of issues, hard to debug and have the big picture. This commit introduces a complete rework based on all knowledge we gained over time.
- Refactored the MongoDB dashboard JSON configuration to enhance structure and readability.
- Introduced a new Python script for generating MongoDB metrics using grafanalib, including common thresholds and state mappings.
- Updated panel configurations and added new metrics for better monitoring capabilities.

Modernize dashboard logic

- Use statically verifiable types for dev time checks
- Ensure we use a more human readable "job"
- Rename variables
- Improve consistency across panels

Reduce code duplication

Also rename mongodb_instance_type to jobs.
Use variable and not sting when possible.

Issue: ZENKO-5003

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+]
+
+
+class MongoDBMetrics:
```
scalgrafanalib has types for metrics, which allow to statically check the labels are correct, and help make promql expressions simpler : better to use them instead of plain strings

francoisferrand replied:
does not seem solved?
did you create a cleanup ticket for later?

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+
+    @staticmethod
+    def replication_lag():
+        return f'last_over_time(mongodb_mongod_replset_member_replication_lag{{job=~"{JOBS["shard_pattern"]}"}}[$interval]) > 0'
```
should it not use `rate_internal` ?

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+        "textMode": "auto",
+    }
+
+    # Update defaults with any provided kwargs
+    for key, default_value in defaults.items():
+        if key not in kwargs:
+            kwargs[key] = default_value
```
not needed, can either use `|` to merge maps:
```
kwargs = defaults | kwargs
```

or simply expand both variables in the `Stat` call:
```
     Stat(
         title=title,
         dataSource=DATASOURCE,
         format=unit,
         thresholds=thresholds or THRESHOLDS["green_red"],
         targets=targets,
         **defaults,
         **kwargs,
     )
```

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+    if targets is None and expr:
+        targets = [Target(expr=expr, legendFormat="__auto")]
+    elif targets is None:
+        targets = []
```
this could be implemented more simply with an extra `default`, relying on kwargs to override the `targets` if needed:

```
def mongodb_stat(title, expr="", unit="none", thresholds=None, **kwargs):
	defaults = {
	    "targets": [Target(expr=expr, legendFormat="__auto")],
	    "colorMode": "value",
	    "textMode": "auto",
	}
```

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+    }
+
+    # Update defaults with any provided kwargs
+    for key, default_value in defaults.items():
```
same as above

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+        **kwargs,
+    )
+    if width:
+        panel.gridPos = GridPos(h=7, w=width, x=0, y=0)
```
not sure where this is used, by generally best to separate the "definition" of the panel from the way it is layout (in the dashboard, when creating the list of panels) : thus there are dedicated functions in `layout` for this purpose, which allow to set the width as well.

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+    "MongoDB services state",
+    orientation="horizontal",
+    targets=[
+        Target(
+            expr=MongoDBMetrics.up(JOBS["configsvr"]),
+            legendFormat="Config Server up:",
+        ),
+        Target(
+            expr=MongoDBMetrics.up(JOBS["mongos"]),
+            legendFormat="MongoS up:",
+        ),
+        Target(
+            expr=MongoDBMetrics.up(JOBS["shards"]),
+            legendFormat="Shard Server up:",
+        ),
```
using python generator helps keep the code safe and simple in such case:
```suggestion
        Target(
            expr=MongoDBMetrics.up(JOBS[job]),
            legendFormat=f`{title} up:`,
        ) for job, title in {
            "configsvr": "Config Server",
            "mongos": "MongoS",
            "shards": "Shard Server", 
        } 
```

(and this may be further improved by adding the "title" in JOBS as well)

francoisferrand replied:
this approach can be used in quite a few places in here to avoid redundancy and make code more expressive (for targets, but maybe for some panels which are very similar)

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+    colorMode="none",
+    orientation="horizontal",
+    textMode="auto",
+    reduceOptions={"calcs": ["lastNotNull"], "fields": "/^mongodb$/", "values": False},
```
> "fields": "/^mongodb$/",

kind of unusual, what is the purpose of this reduce option? isn't this done directly through the prom query ?

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+                            height=8,
+                        ),
+                        # Right side - Shard distribution pie charts horizontally (9 width each)
+                        *layout.resize([docs_distribution_pie], width=9),
+                        *layout.resize([index_size_distribution_pie], width=9),
```
size should not be needed in such case, `layout.row()` is expected to automatically distribute the space on the panels in the row (except for panels which already have a size, obviously)

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+                        # Left side - Config server data stacked (6 width)
+                        layout.column(
+                            [
+                                *layout.resize([docs_configsvr], height=4),
+                                *layout.resize([index_size_configsvr], height=4),
```
resize should not be needed, `layout.column` should do it automatically (I think) by distributing the total size (in param) across the panels

(then it should not be needed either to set the height of the column either, as it is set by the `layout.row`.... but I don't remember to which point the layout engine was polished to cascade all changes :-/ )

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+                        shard_balancing,
+                    ],
+                    height=3,
+                ),
```
a full line for this seems overkill, would this fit in column with the number of indexes panel, in the top 2 rows?

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+        Target(
+            expr=f'{MongoDBMetrics.db_stats("fsTotalSize")} - {MongoDBMetrics.db_stats("fsUsedSize")}',
+            legendFormat="Free",
+        ),
```
this is approximate, since we compare the `max` from each metric: so the total may be more than the available space (if the repartition is not the same on every server)

generally I think this may be acceptable, but it could also create artifacts that can hit us back :-/
maybe we could instead filter on the "primary" of each shard, and/or do the sum() over every instance?

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+            legendFormat="Indexes",
+        ),
+        Target(
+            expr=f'{MongoDBMetrics.db_stats("fsUsedSize")} - {MongoDBMetrics.db_stats("indexSize")}',
```
not correct: the `max` currently done in `db_stats` should be done only after aggregating the 2 series, so that the aggregated series come from the same instance:

```
max(fsUsedSize - indexSize)
```

francoisferrand replied:
still does not seem resolved (esp. if/when we have multiple shards) : to get correct result, we need to combine the metrics first -so that labels are matched-, then call any function which removes the labels. Maybe something like:

```
            expr=f'max(  
                mongodb_dbstats_fsUsedSize{{rs_nm=~"$rs_nm", rs_state="{RS_STATE_PRIMARY}", namespace="{NAMESPACE}"}}
                -
                mongodb_dbstats_indexSize{{rs_nm=~"$rs_nm", rs_state="{RS_STATE_PRIMARY}", namespace="{NAMESPACE}"}}
            )',
```

(probably can't have `max` for both "used" and "free" space; so maybe we should either use sum/avg on both queries, or use `max(used)` and `min(free)`, though with some imprecision if there is some discrepancy in volume usage... ?)

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+    "Average document size",
+    expr=f'avg(mongodb_dbstats_avgObjSize{{database!~"admin|config|local", rs_state="1", job="{JOBS["shards"]}"}})',
+    unit=UNITS.BYTES,
+    decimals=3,
```
do we really need so many decimals? in the screenshot, 138.969 bytes would be better rounded simply to 140

Grafana will automatically convert to the most appropriate metric prefix (k, ...), may be enough to keep 0 or 1 decimal for the average document size?

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+
+num_docs_per_shard = mongodb_stat(
+    "Number of documents per shard",
+    expr='mongodb_dbstats_objects{rs_state="1", database!~"admin|config|local", rs_nm=~"$rs_nm"}',
```
to improve readability, would be good to introduce constant for `rs_state` values : so it is easier to understand what we filter

(also, default filter for `database` and `rs` can be set in the definition of the metrics, when using scalgrafanalib's `*Metrics` classes, to avoid the extra cognitive load here)

francoisferrand replied:
missing a `job` filter here (and namespace I guess), this could pickup stats from any other instance of mongo running in the cluster

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+
+index_size = mongodb_stat(
+    "Index size",
+    expr='max by(rs_nm) (mongodb_dbstats_indexSize{rs_state="1", database!~"admin|config|local", rs_nm=~"$rs_nm"})',
```
missing job label filter

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+    description="MongoDB stores documents in collections. Collections are analogous to tables in relational databases.",
+    targets=[
+        Target(
+            expr=f'max(mongodb_dbstats_storageSize{{database!~"admin|config|local", job="{JOBS["shards"]}"}})',
```
we currently have a single database, so may not be an issue ;

however, since we are filtering on database, I imagine mongo reports separate metrics for each database, right?
And so we need to make the sum over the databases?

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+# Activity indicators for overview
+current_operations = mongodb_stat(
+    "Current Operations/sec",
+    expr=f'sum(rate(mongodb_ss_opcounters{{job=~"{JOBS["all_shards"]}", rs_nm=~"$rs_nm"}}[5m]))',
```
time range should be `$rate_interval`, to scale appropriately with the currently selected time frame?

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+    decimals=1,
+    colorMode="value",
+    thresholds=[Threshold("blue", 0, 0.0)],
+    noValue="0",
```
no data is not **_no_** operation, it can also be no metrics ; and it can be useful to make the distinction.

I don't remember how we handle this in other dashboards: in some case we have a `-` I think, in some case `0` as you did, and probably other things to... may be worth looking for consistency and inspiration, for exemple at cloudserver (which also has "ops/sec" panels)

williamlardier replied:
We use both in cloudserver, but mostly `-`, which I will use for consistency.

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+NAMESPACE = "zenko"
+
+# Job definitions
+JOBS = {
```
using a map like this does not allow the compiler to detect invalid variable access: would be better to use actual fields, and access it with something like `JOBS.configsvr` (which also has less punctuation)

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+JOBS = {
+    "configsvr": "zenko/data-db-mongodb-sharded-configsvr",
+    "mongos": "zenko/data-db-mongodb-sharded-mongos",
+    "shards": "zenko/data-db-mongodb-sharded-shards-data",
+    "shard_pattern": "zenko/data-db-mongodb-sharded-shard.*-data",
+    "all_shards": ".*shards.*",
```
I don't understand the 3 kinds of job for shards : `job` is used to filter based on the prometheus job which performed the scrapping (which maps to the service/servicemonitor in k8s), so either we should have only one way to identify metrics from shard servers... unless this is used accross different metrics, and it should probably be in separate variables/arrays/... to make it clear what the values relate to.

francoisferrand replied:
I see the code is outdated, but no change: some things were not pushed?

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+}
+
+# MongoDB State Mappings for StateTimeline - create a single StatValueMappings with all items
+MONGODB_STATES = [
```
nit: `MONGODB_STATES_MAPPINGS`

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+
+    @staticmethod
+    def version_info():
+        return "avg by (service_name,mongodb) (mongodb_version_info{})"
```
* `avg` may lead to strange results if the versions do not match, I guess? Not sure if we can find a query to return the version is there is only one (there is no "distincts values" function, but maybe comparing min() and max() could do the trick), and otherwise display some sort of error (like "Multiple versions")
* we should filter by job/namespace, to only have the instance(s) we want to look at

williamlardier replied:
I will use

```py
        return f'count by (mongodb) (mongodb_version_info{{job=~"{Jobs.shards}|{Jobs.configsvr}|{Jobs.mongos}", namespace="{NAMESPACE}"}}) > 0'
```

So it returns a real version. Maybe if we `min` it we can display the smallest and thus see if one upgrade did not complete properly?

francoisferrand replied:
> see if one upgrade did not complete properly

To see if the upgrade did not complete, we would need to check if the version is the "expected" one : which may not be the case even if every mongo has the same version.
And to complicate matter, "Zenko" does not know which version of mongo should be running, as today this is a separate solution... (which by the way is -temporarily- normal as well during upgrades, and normally dealt with simply by the declarative nature of operators)
So for now I'd just stick to display/reporting when the version is inconsistent.

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1397 @@
+
+
+class MongoDBMetrics:
+    """MongoDB metrics expressions"""
```
adding this here, as it applies to many metrics : we should properly filter the "cluster" of mongo, as there may be multiple in the cluster → should be done based on the params passed by Zenko (and possibly adding template params, typically in case a single Zenko instance deploys multiple MongoDB clusters - but we are not there yet AFAIK)

this can be done easily using with the Metrics classes, as this can be set as default filters in the definition

francoisferrand replied:
you're absolutely right, we do not need to support multiple clusters yet.

my point was just that we should start (today) with a simple filtering (supporting the single cluster we have now):
```
   VERSION_INFO=Metrics('mongodb_version_info', 'job=~"{Jobs.shards}|{Jobs.configsvr}|{Jobs.mongos}', namespace=namspace)
```

and this can **later** be extended to support multiple mongo clusters.

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+        return f"""min(
+  mongodb_mongod_replset_oplog_head_timestamp{{job=~"{Jobs.shard_pattern}", namespace="{NAMESPACE}"}}
+  -
+  mongodb_mongod_replset_oplog_tail_timestamp{{job=~"{Jobs.shard_pattern}", namespace="{NAMESPACE}"}}
```
here we do the `min` both accross shards AND replicas of a shard.
just to confirm: is this what we min, or should we consider the window for each shard?

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+        return f'last_over_time(mongodb_mongod_replset_member_replication_lag{{job=~"{Jobs.shard_pattern}", namespace="{NAMESPACE}"}}[$__rate_interval]) > 0'
+
+
+def mongodb_stat(title, expr="", unit="none", thresholds=None, **kwargs):
```
best to keep the underlying field name, to avoid confusion?
```suggestion
def mongodb_stat(title, expr="", format="none", thresholds=None, **kwargs):
```

....and then it may be removed from the prototype, and just rely on `kwargs` (and possibly a `defaults` value)

francoisferrand replied:
same for thresholds

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    )
+
+
+def mongodb_timeseries(title, targets, unit="none", thresholds=None, **kwargs):
```
same

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    unit="short",
+    targets=[
+        Target(
+            expr=f'sum by (rs_nm) (mongodb_dbstats_objects{{{DATABASE_FILTER_USER}, rs_state="{RS_STATE_PRIMARY}", rs_nm!~".*configsvr.*", job="{Jobs.shards}", namespace="{NAMESPACE}"}})',
```
is the `rs_nm` filter not redundant with the `job="{Jobs.shards}"` one?

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    unit=UNITS.BYTES,
+    targets=[
+        Target(
+            expr=f'sum by (rs_nm) (mongodb_dbstats_indexSize{{{DATABASE_FILTER_USER}, rs_state="{RS_STATE_PRIMARY}", rs_nm!~".*configsvr.*", job="{Jobs.shards}", namespace="{NAMESPACE}"}})',
```
same

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    "Active cursors",
+    [
+        Target(
+            expr=f'max(mongodb_ss_metrics_cursor_open{{rs_nm=~"$rs_nm", namespace="{NAMESPACE}"}})',
```
missing `jobs` filter

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    "Cursor created",
+    [
+        Target(
+            expr=f'avg by (pod) (increase(mongodb_ss_metrics_cursor_totalOpened{{rs_nm=~"$rs_nm", namespace="{NAMESPACE}"}}[$__rate_interval]))',
```
missing job filter

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    "Disk Reads Completed",
+    [
+        Target(
+            expr=f'sum by(pod) (rate(container_fs_reads_total{{namespace="{NAMESPACE}", pod=~".*mongodb.*"}}[$__rate_interval]))',
```
we should be able to filter on `job` here as well (instead of pod)

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    "Disk Writes Completed",
+    [
+        Target(
+            expr=f'sum by(pod) (rate(container_fs_writes_total{{namespace="{NAMESPACE}", pod=~".*mongodb.*"}}[$__rate_interval]))',
```
same

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    "Average Operation Latency",
+    [
+        Target(
+            expr=f'sum by(type) (\n  rate(mongodb_mongod_op_latencies_latency_total{{job=~"{Jobs.shard_pattern}"}}[$__rate_interval])\n) /\nsum by(type) (\n  rate(mongodb_mongod_op_latencies_ops_total{{job=~"{Jobs.shard_pattern}"}}[$__rate_interval])\n)',
```
should be split/wrapped, for better readability

francoisferrand replied:
nit: also maybe `avg( rate(mongodb_mongod_op_latencies_latency_total[$rate_interval]) / rate(mongodb_mongod_op_latencies_ops_total[$rate_interval]) ) by(type)` may be easier to read (to me at least, I had to think twice to be confident what you wrote gave the expected result indeed); or is there a reason for the slightly more complex query?

williamlardier replied:
It's because we want to compute the weighted average, while in your case, we compute the unweighted one. The first gives a true average latency cluster-wide while your computes the average per shard. I do not want to be shard-specific as I want here to display the right value cluster-wise to have the complete picture. Both are valid though...

francoisferrand replied:
ok (maybe a comment would help), however please also review the others: the PR alternates between both approaches, let's make sure nothing unexpected slips through

williamlardier replied:
I added a comment, and all metrics related to what the customer is experiencing is now using weighted averages.

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    [
+        Target(
+            expr=MongoDBMetrics.replication_lag(),
+            legendFormat="{{pod}} <- {{name}}",
```
`name` is very long, would be best to shorten it somehow.
also not very clear what the "arrow" represents : it may be better to simply report the lag of each (secondary) pod, which will thus be implicetely relative to the primary instance of that shard --> as label, just use the pod/job/name which tracks the (secondary) while lag is being measured

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    showPoints="always",
+    decimals=2,
+    valueMin=0,
+    description="MongoDB replication lag occurs when the secondary node cannot replicate data fast enough to keep up with the rate that data is being written to the primary node.",
```
Is lag not always there, to some extent?
so we may tune this down a bit, to "formally" define lag (something like "lag is the delay needed to replicate data from primary to secondary"), then explain what is expected (should stay within a few seconds) and when to get worried (if lag keeps increasing, it means data cannot be replicated to secondary nodes fast enough to keep up with the rate that data is being written to the primary node)

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+            legendFormat="fc_time",
+        ),
+    ],
+    unit="µs",
```
`fc_count` unit is not mirosecond? So we should have 2 scales in this panel, one for each serie.

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+            legendFormat="{{op_type}}",
+        )
+    ],
+    unit="µs",
```
nit: I see this MICROSECOND unit multiple times, which is unfortunately not in grafanalib; so may be better to add a variable `UNIT_MICROSECOND` ?

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    "Queued Operations",
+    [
+        Target(
+            expr=f'max by (type) (mongodb_mongod_global_lock_current_queue{{job=~"{Jobs.shard_pattern}", rs_nm=~"$rs_nm"}})',
+            legendFormat="{{type}}",
+        )
+    ],
+    unit=UNITS.OPS_PER_SEC,
```
looking at the metrics (`mongodb_mongod_global_lock_current_queue`), it seems weird that the unit would be `ops/sec`, as it looks (from the name "current") and is used (wihtout rate) like a gauge... Is the unit wrong, and should just the a number?

francoisferrand replied:
also, it is really a `max` ?
- we should filter only primary instances (of each shard)
- when multiple shards, we should `sum` the locks of each shard

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    "Operation queue size due to locks",
+    [
+        Target(
+            expr=f'sum by(job) (mongodb_ss_globalLock_currentQueue{{count_type=~"readers|writers", job=~"{Jobs.shard_pattern}", rs_nm=~"$rs_nm"}})',
```
looks pretty similar to the previous panel's metric (`mongodb_ss_globalLock_currentQueue` vs `mongodb_mongod_global_lock_current_queue`, except for the "by type") : what is the difference? how do they relate?
If there is some aggregation (implicit, on mongo side), may be better to make this clear in the title: maybe something like "Queued operations by type" and "Total queued operations"

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    "Write Conflicts per Second",
+    [
+        Target(
+            expr=f'sum by(job) (\n  rate(mongodb_mongod_metrics_operation_total{{job=~"{Jobs.shard_pattern}",state="writeConflicts"}}[$__rate_interval])\n) / sum by(job) (\n  rate(mongodb_ss_opcounters{{job=~"{Jobs.shard_pattern}", legacy_op_type=~"insert|update|delete"}}[$__rate_interval])\n)',
```
wrap and use `avg` to make this easier to read?

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    [
+        Target(
+            expr=f'sum by(job) (\n  rate(mongodb_mongod_metrics_operation_total{{job=~"{Jobs.shard_pattern}",state="writeConflicts"}}[$__rate_interval])\n) / sum by(job) (\n  rate(mongodb_ss_opcounters{{job=~"{Jobs.shard_pattern}", legacy_op_type=~"insert|update|delete"}}[$__rate_interval])\n)',
+            legendFormat="Write conflict",
```
not correct, same legend for each job

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    "WiredTiger Engine Cache hit/miss ratio",
+    [
+        Target(
+            expr=f'(\n   rate(mongodb_ss_wt_cache_pages_requested_from_the_cache{{rs_state="{RS_STATE_PRIMARY}",rs_nm=~"$rs_nm", namespace="{NAMESPACE}"}}[$__rate_interval]) - \n   rate(mongodb_ss_wt_cache_pages_read_into_cache{{rs_state="{RS_STATE_PRIMARY}",rs_nm=~"$rs_nm", namespace="{NAMESPACE}"}}[$__rate_interval])) / \n   rate(mongodb_ss_wt_cache_pages_requested_from_the_cache{{rs_state="{RS ...
```
Should be simplified (and wrapped), something like:
```suggestion
            expr=f'1 -  rate(mongodb_ss_wt_cache_pages_read_into_cache{{rs_state="{RS_STATE_PRIMARY}",rs_nm=~"$rs_nm", namespace="{NAMESPACE}"}}[$__rate_interval])
                        /
                        rate(mongodb_ss_wt_cache_pages_requested_from_the_cache{{rs_state="{RS_STATE_PRIMARY}",rs_nm=~"$rs_nm", namespace="{NAMESPACE}"}}[$__rate_interval])',
```

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    "Stacked CPU usage",
+    [
+        Target(
+            expr=f'sum by (rs_nm) (\n  (\n    rate(mongodb_sys_cpu_{cpu_type}_ms{{job="{Jobs.shards}", rs_nm=~"$rs_nm"}}[$__rate_interval])\n  ) / 1000\n)',
```
missing namespace filter

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+        ),
+        Target(
+            expr=f'sum(irate(mongodb_mongod_metrics_query_executor_total{{job=~"{Jobs.shard_pattern}", state="scanned"}}[5m])) /\nsum(irate(mongodb_mongod_metrics_document_total{{job=~"{Jobs.shard_pattern}", state="returned"}}[5m]))',
+            legendFormat="Scanned idx / returned",
```
missing namespace filter

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    "Memory Usage",
+    [
+        Target(
+            expr=f'mongodb_ss_mem_resident{{pod=~".*mongodb.*", namespace="{NAMESPACE}", rs_nm=~"$rs_nm"}}',
```
should filter on `job` instead of `pod`

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    "Network I/O",
+    [
+        Target(
+            expr=f'{prefix}rate(mongodb_ss_network_bytes{direction}{{pod=~".*mongodb.*", namespace="{NAMESPACE}", rs_nm=~"$rs_nm"}}[$__rate_interval])',
```
same

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    "Disk IO Utilization (all instances)",
+    [
+        Target(
+            expr=f'sum by(pod) (rate(container_blkio_device_usage_total{{operation="Total", namespace="{NAMESPACE}", pod=~".*mongodb.*"}}[$__rate_interval]))',
```
same

williamlardier replied:
Must keep pod here, as this is not a mongodb exported metric... I will define and use `POD_PATTERN` in several places, for these metrics. I think it is valuable to keep these panels in this dashboard, even if not directly exported by mongodb, because it gives the big picture that we are often missing in current dashboards... But please challenge if you disagree.

francoisferrand replied:
should not be a `POD_PATTERN` (hardcoded in the dashboard code), but rely on inputs specified by ZKOP.

(or create a -hidden- template parameter which extracts the values of `pod` label from the `up{job=...}` query, to get this list)

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    "Available Concurrency Tickets",
+    [
+        Target(
+            expr=f'avg(rate(mongodb_mongod_wiredtiger_concurrent_transactions_total_tickets{{txn_rw="{txn_type}", namespace="{NAMESPACE}"}}[$__rate_interval]) - rate(mongodb_mongod_wiredtiger_concurrent_transactions_out_tickets{{txn_rw="{txn_type}", namespace="{NAMESPACE}"}}[$__rate_interval]))',
```
missing job filter

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+        )
+        for txn_type in ["read", "write"]
+    ],
+    unit="none",
```
Unit seem wrong : the measure is the different between the *rate* of total and out "concurrent transactions tickets", so it is still a *rate*.
It also does not seem to match the title : "available currency tickets" should be something like the max/total [and likely fixed --> rate() == 0] number of tickets minus the "currently in-use" concurrency tickets.

francoisferrand replied:
`mongodb_mongod_wiredtiger_concurrent_transactions_available_tickets` would seem a better metric for this panel...

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    "Write Concurrency Tickets",
+    [
+        Target(
+            expr=f'sum by (txn_rw) (mongodb_mongod_wiredtiger_concurrent_transactions_available_tickets{{job=~"{Jobs.shard_pattern}", txn_rw="write"}})',
+            legendFormat="{{txn_rw}}",
+        )
```
should be "by pod" ?

```suggestion
            expr=f'sum by (pod) (mongodb_mongod_wiredtiger_concurrent_transactions_available_tickets{{job=~"{Jobs.shard_pattern}", txn_rw="write"}})',
            legendFormat="{{pod}}",
        )
```

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+    ],
+    fillOpacity=0,
+    showPoints="auto",
+    legendDisplayMode="table",
+    legendPlacement="right",
+    legendValues=["mean", "max", "min"],
+)
+
+read_concurrency_tickets = mongodb_timeseries(
+    "Read Concurrency Tickets",
+    [
```
nit: both panels are the same, only different by read/write:


```suggestion
write_concurrency_tickets, read_concurrency_tickets = [
    mongodb_timeseries(
        f'${txn_rw} concurrency tickets',
        ...
        Target(
            expr=f'sum by (pod) (mongodb_mongod_wiredtiger_concurrent_transactions_available_tickets{{job=~"{Jobs.shard_pattern}", txn_rw="${txn_rw}"}})',
            legendFormat="{{pod}}",
        )
    )
    for txn_rw in ['write', 'read']
]
```

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+            expr=f'rate(mongodb_ss_wt_cache_modified_pages_evicted_by_application_threads{{job=~"{Jobs.shard_pattern}", rs_state="{RS_STATE_PRIMARY}", rs_nm=~"$rs_nm"}}[$__rate_interval])',
+            legendFormat="{{pod}}",
+        )
+    ],
```
missing namespace filter

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+        ],
+        templating=Templating(
+            [
+                Template(
+                    name="interval",
+                    label="Interval",
+                    query="1s,5s,1m,5m,1h,6h,1d",
+                    auto=True,
+                    autoCount=200,
+                    autoMin="1s",
+                    includeAll=False,
+                    multi=False,
+                    type="interval",
+                ),
```
not used?

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+                    dataSource=DATASOURCE,
+                    name="rs_nm",
+                    label="Instance",
+                    query="label_values(mongodb_mongod_replset_my_state,set)",
```
should use the job and namespace filters

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+                label="Prometheus",
+                pluginId="prometheus",
+                pluginName="Prometheus",
+            ),
```
namespace and job (or prefix) should be passed as inputs (and they will be set by ZKOP)

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1359 @@
+                    dataSource=DATASOURCE,
+                    name="node_name",
+                    label="Node Name",
+                    query='label_values(mongodb_up{cluster=~"$cluster",service_name=~"$service_name"}, node_name)',
```
`cluster` and `service_name` are neither inputs nor "earlier" template --> will not work.
Anyway this template param does not seem to be used (`node_name` is not used in the file, and the template is hidden)

ghost commented:
@francoisferrand : The PR mesage was updated with the new dashboard image. I checked one by one and I resolved all comments, plus some related. Just one thing, I changed (removed) the Jobs class to just use it as an input from the dashboard:

<img width="434" height="203" alt="image" src="https://github.com/user-attachments/assets/4bca2253-e12d-4b7d-9ec6-fa1b7fe5a564" />

Please tell me if you had something else in mind.

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1353 @@
+            [
+                Template(
+                    dataSource=DATASOURCE,
+                    name="job",
```
this has the same name as the `input`, which makes it quite confusing

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1353 @@
+                    dataSource=DATASOURCE,
+                    name="job",
+                    label="Job",
+                    query='label_values(mongodb_up{namespace="zenko"},job)',
```
this should use the namespace and job inputs to filter the returned jobs:

```suggestion
                    query='label_values(mongodb_up{namespace="${namespace}", job=~"${job}-.*"},job)',
```

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1353 @@
+            [
+                RowPanel(
+                    title="Overview",
+                    repeat=Repeat(variable="service_name", direction="h"),
```
why are we repeating this?

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1353 @@
+    "Cache Size",
+    [
+        Target(
+            expr=f'sum({metric}{{job=~"$job", namespace="{NAMESPACE}"}}) by ()',
```
nit:
```suggestion
            expr=f'sum({metric}{{job=~"$job", namespace="{NAMESPACE}"}})',
```

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1353 @@
+    [
+        Target(
+            expr=f'mongodb_mongod_replset_oplog_head_timestamp{{job=~"$job", namespace="{NAMESPACE}"}}-mongodb_mongod_replset_oplog_tail_timestamp{{job=~"$job", namespace="{NAMESPACE}"}}',
+            legendFormat="{{ job }}",
```
some panels are using `job` as label (i.e. `zenko/artesca-data-mongo-sharded-shard0-data1`) while others are using `pod` (i.e. `artesca-data-mongo-sharded-shard0-data1`) : should make things consistent, and choose one (`pod` seems more readable, IMHO)

williamlardier replied:
Side note: by reworking the way we work with the "job" we can simplify this, and at the same time, unify it...

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1353 @@
+
+# Constants
+DATASOURCE = "${DS_PROMETHEUS}"
+NAMESPACE = "zenko"
```
Need to use the template variable:

```suggestion
NAMESPACE = "${namespace}"
```

(we may also inline this, not sure the variable helps readability)

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1353 @@
+                    includeAll=True,
+                    multi=True,
+                    refresh=1,
+                    type="query",
```
the `zenko/` prefix does not look very good in the selector, we may set the regex/pattern to improve the display by removing the `${namespace}/` prefix

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1353 @@
+        return f'count by (mongodb) (mongodb_version_info{{job=~"$job", namespace="{NAMESPACE}"}}) > 0'
+
+    @staticmethod
+    def db_stats(metric, filters=""):
```
not used anymore

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1353 @@
+)
+
+avg_doc_size = mongodb_stat(
+    "Average doc. size",
```
Somehow looks better with just "document", or at least removing "shortened" dot indicator...
If we really need to shorten this, I'd better go with "Avg document size" or "Avg doc size" ?

francoisferrand replied:
(we should probably ask TechPub for advice here)

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1353 @@
+)
+
+num_docs_per_shard = mongodb_stat(
+    "Number of doc./shard",
```
same, I think "Avg documents per shard" or "Avg docs per shard" seems better IMHO

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1353 @@
+
+# Activity indicators for overview
+current_operations = mongodb_stat(
+    "Current Op/sec",
```
"Operations rate" or even just "Operations" ?
(the unit is actually displayed below, no need to repeat...)

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1353 @@
+    "Data operations",
+    [
+        Target(
+            expr=MongoDBMetrics.opcounters("query|getmore"),
+            legendFormat="Read - {{ job }}",
+        ),
+        Target(
+            expr=MongoDBMetrics.opcounters("insert|update|delete"),
+            legendFormat="Write - {{ job }}",
```
would these not make more sense by `shard`, i.e. use `shard` in metrics, and sum by(shard) in `opcounters` ?

williamlardier replied:
`shard` is not defined/a valid metric for mongodb_ss_opcounters. It also does not make much sense because we are looking at the mongos level, which is abstracting the shards. So I would say that no, it does not makes really sense. I don't think we'll need that metric for a specific shard either, and that will stay visible because the job itself will be different per shard...

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1353 @@
+    [
+        Target(
+            expr=f'avg(sum by(job) (rate(mongodb_mongod_metrics_operation_total{{job=~"$job", namespace="{NAMESPACE}", state="writeConflicts"}}[$__rate_interval])) / sum by(job) (rate(mongodb_ss_opcounters{{job=~"$job", namespace="{NAMESPACE}", legacy_op_type=~"insert|update|delete"}}[$__rate_interval])))',
+            legendFormat="{{job}}",
```
since there is an `avg()`, there won't be any `job` label anymore

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1353 @@
+    ],
+    unit=UNITS.PERCENT_FORMAT,
+    thresholds=THRESHOLDS["write_conflicts"],
+    fillOpacity=27,
```
is the 27 expected? (all -or most- others are at 20)

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1353 @@
+
+# Additional Performance Panels
+op_queue_size = mongodb_timeseries(
+    "Total Queued Operations by Job",
```
`by pod` or `by instance` would be more relevant (`job` is our implementation detail, not a "business" concept with mongo)

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1353 @@
+    "Ratio of documents scanned per query",
+    [
+        Target(
+            expr=f'sum(irate(mongodb_mongod_metrics_query_executor_total{{job=~"$job", namespace="{NAMESPACE}", state="scanned_objects"}}[5m])) / sum(irate(mongodb_mongod_metrics_document_total{{job=~"$job", namespace="{NAMESPACE}", state="returned"}}[5m]))',
```
```suggestion
            expr=f'sum(irate(mongodb_mongod_metrics_query_executor_total{{job=~"$job", namespace="{NAMESPACE}", state="scanned_objects"}}[$__rate_interval])) / sum(irate(mongodb_mongod_metrics_document_total{{job=~"$job", namespace="{NAMESPACE}", state="returned"}}[$__rate_interval]))',
```

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1353 @@
+            legendFormat="Scanned objects / returned",
+        ),
+        Target(
+            expr=f'sum(irate(mongodb_mongod_metrics_query_executor_total{{job=~"$job", namespace="{NAMESPACE}", state="scanned"}}[5m])) / sum(irate(mongodb_mongod_metrics_document_total{{job=~"$job", namespace="{NAMESPACE}", state="returned"}}[5m]))',
```
```suggestion
            expr=f'sum(irate(mongodb_mongod_metrics_query_executor_total{{job=~"$job", namespace="{NAMESPACE}", state="scanned"}}[$__rate_interval])) / sum(irate(mongodb_mongod_metrics_document_total{{job=~"$job", namespace="{NAMESPACE}", state="returned"}}[$__rate_interval]))',
```

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1353 @@
+            expr=f'sum by (pod) (mongodb_mongod_wiredtiger_concurrent_transactions_available_tickets{{txn_rw="{txn_type}", job=~"$job", namespace="{NAMESPACE}"}})',
+            legendFormat=f"{{{{pod}}}} {txn_type}",
+        )
+        for txn_type in ["read", "write"]
```
```suggestion
        Target(
            expr=f'sum by (pod, txn_type) (mongodb_mongod_wiredtiger_concurrent_transactions_available_tickets{{txn_rw="read|write", job=~"$job", namespace="{NAMESPACE}"}})',
            legendFormat=f"{{{{pod}}}} {txn_type}",
        )
```
(and we may remove the filter on `txn_type`, if there are only 2 values)

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -0,0 +1,1353 @@
+    [
+        Target(
+            expr=f'sum by (pod) (mongodb_mongod_wiredtiger_concurrent_transactions_available_tickets{{txn_rw="{txn_type}", job=~"$job", namespace="{NAMESPACE}"}})',
+            legendFormat=f"{{{{pod}}}} {txn_type}",
```
txn_type is more redundant & predictable (short!), so put it first ; and add a separator?
```suggestion
            legendFormat=f"{txn_type} - {{{{pod}}}}",
```

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -1062,43 +1446,51 @@ def mongodb_state_timeline(title, expr, description="", mappings=None, **kwargs)
-            ("mongodb_mongod_wiredtiger_cache_bytes", "used"),
-            ("mongodb_ss_wt_cache_tracked_dirty_bytes_in_the_cache", "dirty"),
-        ]
+            expr="sum(" + Metrics.MONGOD_WIREDTIGER_CACHE_MAX_BYTES() + ")",
+            legendFormat="max",
+        ),
+        Target(
+            expr="sum(" + Metrics.MONGOD_WIREDTIGER_CACHE_BYTES() + ")",
+            legendFormat="used",
+        ),
+        Target(
+            expr="sum(" + Metrics.SS_WT_CACHE_TRACKED_DIRTY_BYTES_IN_THE_CACHE() + ")",
+            legendFormat="dirty",
+        ),
```
nit: why not use generation?

```suggestion
            expr="sum(" + metric() + ")",
            legendFormat=legend,
        )
        for metric, legend in [
            (Metrics.MONGOD_WIREDTIGER_CACHE_MAX_BYTES, "max"),
            (Metrics.MONGOD_WIREDTIGER_CACHE_BYTES, "used"),
            (Metrics.SS_WT_CACHE_TRACKED_DIRTY_BYTES_IN_THE_CACHE, "dirty"),
        ] 
```

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -1136,9 +1528,10 @@ def mongodb_state_timeline(title, expr, description="", mappings=None, **kwargs)
-                    query='label_values(mongodb_up{namespace="zenko"},job)',
+                    name="mongodb_instance_type",
+                    label="MongoDB instance type",
+                    query='label_values(mongodb_up{namespace="${namespace}"}, job)',
```
Should still rely on the `job` template parameter provided by ZKOP, to ensure we return only the expected instance of mongo:

```suggestion
                    query='label_values(mongodb_up{namespace="${namespace}", job=~"^${job}"}, job)',
```

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -1136,9 +1528,10 @@ def mongodb_state_timeline(title, expr, description="", mappings=None, **kwargs)
-                    name="job",
-                    label="Job",
-                    query='label_values(mongodb_up{namespace="zenko"},job)',
+                    name="mongodb_instance_type",
```
nit: `job=~"$mongodb_instance_type"` looks kind of weird → we may use `jobs` here, which is not the same as the template `job` param, and conveys the fact that this may contain more than one job (and should thus be matched with =~)

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -277,31 +551,57 @@ def mongodb_state_timeline(title, expr, description="", mappings=None, **kwargs)
-            expr=f'sum(mongodb_dbstats_indexSize{{{DATABASE_FILTER_USER}, job=~"$job", rs_state="{RS_STATE_PRIMARY}", namespace="{NAMESPACE}"}})',
+            expr="sum("
+            + Metrics.DBSTATS_INDEX_SIZE(
+                DATABASE_FILTER_USER, f'rs_state="{RS_STATE_PRIMARY}"'
```
parameter should be passed as `kwargs` when possible (may be others like this):
```suggestion
                DATABASE_FILTER_USER, rs_state=RS_STATE_PRIMARY
```

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -277,31 +551,57 @@ def mongodb_state_timeline(title, expr, description="", mappings=None, **kwargs)
-            expr=f'sum(mongodb_dbstats_fsUsedSize{{{DATABASE_FILTER_USER}, job=~"$job", rs_state="{RS_STATE_PRIMARY}", namespace="{NAMESPACE}"}} - mongodb_dbstats_indexSize{{{DATABASE_FILTER_USER}, job=~"$job", rs_state="{RS_STATE_PRIMARY}", namespace="{NAMESPACE}"}})',
+            expr="sum("
+            + Metrics.DBSTATS_FS_USED_SIZE(
+                DATABASE_FILTER_USER, f'rs_state="{RS_STATE_PRIMARY}"'
```
as we set `rs_state` to `RS_STATE_PRIMARY` most of the type, maybe we should set it by default to this value (directly in metrics parameter definition, like `namespace`)

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -314,7 +614,7 @@ def mongodb_state_timeline(title, expr, description="", mappings=None, **kwargs)
     targets=[
         Target(
-            expr=f'sum(mongodb_dbstats_storageSize{{{DATABASE_FILTER_USER}, job=~"$job", namespace="{NAMESPACE}"}})',
+            expr="sum(" + Metrics.DBSTATS_STORAGE_SIZE(DATABASE_FILTER_USER) + ")",
```
same for database, you should consider if a default value could be appropriate, to simplify code (which does not prevent overriding when a different filter is needed)

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -382,8 +690,12 @@ def mongodb_state_timeline(title, expr, description="", mappings=None, **kwargs)
+                DATABASE_FILTER_USER, f'rs_state="{RS_STATE_PRIMARY}"'
+            )
+            + ")",
+            legendFormat="{{pod}}",
```
inconsistent `sum by (job)` vs `legendFormat="{{pod}}"`

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -399,8 +711,12 @@ def mongodb_state_timeline(title, expr, description="", mappings=None, **kwargs)
+                DATABASE_FILTER_USER, f'rs_state="{RS_STATE_PRIMARY}"'
+            )
+            + ")",
+            legendFormat="{{pod}}",
```
same

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -409,59 +725,75 @@ def mongodb_state_timeline(title, expr, description="", mappings=None, **kwargs)
     "Shard Server States",
-    MongoDBMetrics.replset_state("zenko/data-db-mongodb-sharded-shard.*-data"),
+    "max by (pod) ("
+    + Metrics.REPLSET_STATE('job=~"zenko/data-db-mongodb-sharded-shard.*-data"')
```
job filter(s) should ideally not be hard-coded, but come from ZKOP...
what we often do is that ZKOP sets multiple params, for each kind of job: so in this case it may be job_configsvr, job_shards, ...

Though in this instance, Mongo is deployed by the chart in this repo, so there will always be some coupling :-/ ... So at the very least it should be made consistent:

```suggestion
    + Metrics.REPLSET_STATE('job=~"${namespace}/${job}-shard.*-data"')
```

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -409,59 +725,75 @@ def mongodb_state_timeline(title, expr, description="", mappings=None, **kwargs)
     "Shard configuration States",
-    MongoDBMetrics.replset_state("zenko/data-db-mongodb-sharded-configsvr"),
+    "max by (pod) ("
+    + Metrics.REPLSET_STATE('job=~"zenko/data-db-mongodb-sharded-configsvr"')
```
same

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -486,84 +818,88 @@ def mongodb_state_timeline(title, expr, description="", mappings=None, **kwargs)
         Target(
-            expr=MongoDBMetrics.opcounters("query|getmore"),
+            expr="sum by (job) (rate("
+            + Metrics.SS_OPCOUNTERS('legacy_op_type=~"query|getmore"')
```
nit: you can pass a `list` to specify the list of params with a nicer syntax


```suggestion
            + Metrics.SS_OPCOUNTERS(legacy_op_type=['query', 'getmore'])
```

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -599,44 +938,46 @@ def mongodb_state_timeline(title, expr, description="", mappings=None, **kwargs)
-        for client_type in ["readers", "writers"]
-    ]
-    + [
+            expr="avg(" + Metrics.SS_CONNECTIONS(conn_type="current") + ")",
```
nit: generators from earlier revisions were more readable than the duplicated code (not just here, there are quite a few)

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -763,68 +1104,73 @@ def mongodb_state_timeline(title, expr, description="", mappings=None, **kwargs)
+            expr="sum by(pod) ("
+            + Metrics.SS_GLOBAL_LOCK_CURRENT_QUEUE(f'count_type=~"readers|writers"')
+            + ")",
+            legendFormat="{{pod}}",
```
nit: in such case (e.g. the expression returns a series with a single label), I wonder if Grafana does not actually automatically set the legend to that label... (not an issue, but this could remove qui a few lines)

williamlardier replied:
No, unfortunately: `{job="zenko/data-db-mongodb-sharded-shard0-data"}`

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -143,177 +144,177 @@ class Metrics:
         "rs_state",
         namespace="${namespace}",
-    ).with_defaults(f'job=~"{JOB_PATTERN}"')
+    ).with_defaults(DATABASE_FILTER_USER, JOB_FILTER, rs_state=RS_STATE_PRIMARY)
```
nit: FYI, `with_defaults` is kind of the initial solution to provide default value, but it requires duplicating the name of the field. The preferred approach -when possible- is to use named parameters to provide both the field name and value, like what is done for `namespace`:

```suggestion
        rs_state=RS_STATE_PRIMARY,
        namespace="${namespace}",
    ).with_defaults(DATABASE_FILTER_USER, JOB_FILTER)
```

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -992,7 +936,6 @@ def mongodb_state_timeline(title, expr, description="", mappings=None, **kwargs)
             + Metrics.CONTAINER_FS_READS_TOTAL(f'pod=~"{POD_PATTERN}"')
             + "))",
-            legendFormat="{{pod}}",
         )
```
`legendFormat` removed, seems to contradict you earlier response about this: is this a left-over, or indeed in this specific case (line 995 and 1010) it works without the `{{pod}}` legend?

francoisferrand commented on monitoring/mongodb/dashboard.py:
```diff
@@ -819,7 +778,7 @@ def mongodb_state_timeline(title, expr, description="", mappings=None, **kwargs)
         Target(
             expr="max by (job) (" + Metrics.MONGOS_SHARDING_CHUNKS_IS_BALANCED() + ")",
-            legendFormat="{{pod}}",
+            legendFormat="{{ job }}",
```
there was a comment (earlier) about consistently choosing between `job` and `pod` in legends: at other places with do it `by(pod)`, should the fix here and line 833 not be do do `max by(pod)` ?

```suggestion
expr="max by (pod) (" + Metrics.MONGOS_SHARDING_CHUNKS_IS_BALANCED() + ")",
            legendFormat="{{pod}}",
```

francoisferrand commented:
/after_pull_request 2240

ghost commented:
/approve

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZENKO-5003 contains:


* `2.12.4`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `2.12.5`


Please check the `Fix Version/s` of ZENKO-5003, or the target
branch of this pull request.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.12`



The following branches will **NOT** be impacted:

* `development/2.10`
* `development/2.11`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.12`



The following branches have **NOT** changed:

* `development/2.10`
* `development/2.11`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



Please check the status of the associated issue ZENKO-5003.

Goodbye williamlardier.
